### ISSUE TITLE
Question about `perturb_vec` in the calculation of GANHVPOperator

### ISSUE THREAD

**Issue by Victarry** [Animadversio/GAN-Geometry#5]
Thanks for your great work and user-friendly repo to reproduce results and the notebook of tutorial. It really inspires me.

I am a little confused about the calculation of gradient in `GANHVPOperator`. 

https://github.com/Animadversio/GAN-Geometry/blob/a6e0e555d3c1449a72ca1758e71d4aac3364c03c/core/GAN_hvp_operator.py#L77-L87

**What the purpose of  `self.perturb_vec`** ?

For example, I may instead calculate gradient like this:
```python
self.code = code.requires_grad_(True)
img = self.model.visualize(self.code, )  # forward the feature vector through the GAN 
d_sim = self.criterion(self.preprocess(img), self.preprocess(img.detach())) 
# similarity metric between 2 images. 
gradient = torch.autograd.grad(d_sim, self.perturb_vec, create_graph=True, retain_graph=True)[0] 
```

What's the advantage of using perturb_vec?

**Comment by Animadversio** [comment]
Thanks for your compliment! Nice to know my repo is of interest and useful to you.

Conceptually, `self.perturb_vec` is the small perturbation vector i.e. `dz`
Practically, `self.perturb_vec` was added to avoid the criterion being **exactly zero**. This is important for the forward mode differentiation. 

See for the backward mode HVP, the effect of it is discussable. 
For LPIPS or L2 image distance $d=(\phi(x)-\phi(y))^2$ or $d=\|x-y\|^2$, these distances both has a square in it. 
Then if $\phi(x)==\phi(y)$ i,e, the two inputs are exactly the same then you will get 0 gradients to either x or y, then *I thought* the higher-order gradient may face some problem with zero gradient. To avoid this, I added the `perturb_vector`, dz to perturb the base vector so the first order gradient is non-zero. 

HOWEVER, recently I retried this and found even if the first-order gradient is zero, the `Hvp` product is correct. 
This suggests I may not actually need the perturbation vector and can just compute Hession on the spot! (good catch)

A demo of this for L2 loss. 
```python 
import torch as th
x = th.tensor([1.0,2.0])
y = x.clone().requires_grad_(True)
loss = th.dot(y-x,y-x)
y_grad = th.autograd.grad(loss, y, create_graph=True, retain_graph=True)[0]
vec = th.tensor([2.0,1.0])
y_gradgrad = th.autograd.grad(y_grad, y, grad_outputs=vec, only_inputs=True, retain_graph=True)[0]
```
The Hessian is 2 times Identity matrix, so the `y_gradgrad == 2 * vec` which is correct. 
```python
y_grad
Out: tensor([0., 0.], grad_fn=<AddBackward0>)
y_gradgrad
Out: tensor([4., 2.])
```


Let me know if this didn't answer your question.
